website/docs/configuration/expressions.html.md

```diff
@@ -16,195 +16,204 @@ references to data exported by resources, arithmetic, conditional evaluation,
-so usually object and map values can be used interchangably as long as their
-contained values are of suitable types.
+Expressions are most often used to set values for the arguments of resources and
+child modules. In these cases, the argument has an expected type and the given
+expression must produce a value of that type.
 
-Likewise, Terraform will automatically convert tuple values to list values
-when required, and so tuple and list values can be used interchangably in
-most cases too.
+Where possible, Terraform automatically converts values from one type to
+another in order to produce the expected type. If this isn't possible, Terraform
+will produce a type mismatch error and you must update the configuration with a
+more suitable expression.
 
-Because of these automatic conversions, it is common to not make a strong
-distinction between object and map or tuple and list in everyday discussion
-of the Terraform language. The Terraform documentation usually discusses the
-object and tuple types only in rare cases where it is important to distinguish
-them from the map and list types.
+Terraform automatically converts number and bool values to strings when needed.
+It also converts strings to numbers or bools, as long as the string contains a
+valid representation of a number or bool value.
 
-## References to Named Objects
+* `true` converts to `"true"`, and vice-versa
+* `false` converts to `"false"`, and vice-versa
+* `15` converts to `"15"`, and vice-versa
 
-A number of different named objects can be accessed from Terraform expressions.
-For example, resources are available in expressions as named objects that have
-an object value corresponding to the schema of their resource type, accessed by
-a dot-separated sequence of names like `aws_instance.example`.
+## Literal Expressions
 
-The following named objects are available:
+A _literal expression_ is an expression that directly represents a particular
+constant value. Terraform has a literal expression syntax for each of the value
+types described above:
+
+* Strings are usually represented by a double-quoted sequence of Unicode
+  characters, `"like this"`. There is also a "heredoc" syntax for more complex
+  strings. String literals are the most complex kind of literal expression in
+  Terraform, and have additional documentation on this page:
+    * See [String Literals](#string-literals) below for information about escape
+      sequences and the heredoc syntax.
+    * See [String Templates](#string-templates) below for information about
+      interpolation and template directives.
+* Numbers are represented by unquoted sequences of digits with or without a
+  decimal point, like `15` or `6.283185`.
+* Bools are represented by the unquoted symbols `true` and `false`.
+* The null value is represented by the unquoted symbol `null`.
+* Lists/tuples are represented by a pair of square brackets containing a
+  comma-separated sequence of values, like `["a", 15, true]`.
```

Comment:
Another important spot where I conflated list/tuples etc. a bit more for convenience.